crates/noirc_evaluator/src/ssa/code_gen.rs

```diff
@@ -233,60 +236,69 @@ impl<'a> IRGenerator<'a> {
         env: &mut Environment,
         stmt_id: &StmtId,
-    ) -> Result<(), RuntimeError> {
+    ) -> Result<Value, RuntimeError> {
         let statement = self.def_interner().statement(stmt_id);
         match statement {
             HirStatement::Constrain(constrain_stmt) => {
                 self.handle_constrain_statement(env, constrain_stmt)
             }
             HirStatement::Expression(expr) | HirStatement::Semi(expr) => {
-                self.expression_to_object(env, &expr)?;
-                Ok(())
+                self.expression_to_object(env, &expr)
             }
             HirStatement::Let(let_stmt) => {
                 // let statements are used to declare a higher level object
                 self.handle_let_statement(env, let_stmt)
             }
             HirStatement::Assign(assign_stmt) => {
-                //////////////TODO name is needed because we don't parse main arguments
-                let (ident_def, ident_name) = self.lvalue_ident_def_and_name(&assign_stmt.lvalue);
-
-                let rhs = self.expression_to_object(env, &assign_stmt.expression)?;
-
-                if let Some(lhs) = self.find_variable(ident_def) {
-                    // We may be able to avoid cloning here if we change find_variable
-                    // and assign_pattern to use only fields of self instead of `self` itself.
-                    let lhs = lhs.clone();
-                    let result = self.assign_pattern(&lhs, rhs);
-                    self.variable_values.insert(ident_def, result);
-                } else {
-                    //var is not defined,
-                    //let's do it here for now...TODO
-                    let typ = self.lvalue_type(&assign_stmt.lvalue);
-                    self.bind_fresh_pattern(&ident_name, &typ, rhs);
-                }
-
-                Ok(())
+                self.handle_assign_statement(assign_stmt.lvalue, assign_stmt.expression, env)
             }
             HirStatement::Error => unreachable!(
                 "ice: compiler did not exit before codegen when a statement failed to parse"
             ),
         }
     }
 
-    fn lvalue_type(&self, lvalue: &HirLValue) -> Type {
-        match lvalue {
-            HirLValue::Ident(ident) => self.def_interner().id_type(ident.id),
-            HirLValue::MemberAccess { .. } => unimplemented!(),
-            HirLValue::Index { .. } => unimplemented!(),
+    pub fn single_value(&self, val: &Value) -> NodeId {
+        let nodes = function::value_to_node_id(val);
+        assert!(nodes.len() == 1);
+        nodes[0]
+    }
+
+    fn evaluate_indexed_value(
+        &mut self,
+        array: &HirLValue,
+        index: ExprId,
+        env: &mut Environment,
+    ) -> (u32, NodeId) {
+        let ident_def = self.lvalue_ident_def(array);
+        if let Some(val) = self.find_variable(ident_def) {
+            let lhs = function::value_to_node_id(val);
+            assert!(lhs.len() == 1);
```

Comment:
What in case of array of structure?

Reply:
I'm not suggesting the whole function be replaced, but this block with `function::to_node_id` and asserting that its length is 1 is the same as what the single_value function was doing, so it can be re-used.

Reply:
My point is that for now I assert the length is 1 but in future we will use the to_nodeids version.